umode-api/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+
+impl IntoRegisters for Result<(), Error> {
+    fn from_registers(regs: &[u64]) -> Result<(), Error> {
+        match regs[0] {
+            HYPC_SUCCESS => Ok(()),
+            e => Err(e.into()),
+        }
+    }
+
+    fn set_registers(&self, regs: &mut [u64]) {
+        match self {
+            Ok(_) => {
+                regs[0] = HYPC_SUCCESS;
+            }
+            Err(e) => {
+                regs[0] = *e as u64;
+            }
+        }
+    }
+}
+
+// UmodeRequest: calls from hypervisor to Umode requesting an operation.
+
+/// Umode operations.
+#[derive(Debug, Clone, Copy)]
+#[repr(u64)]
+pub enum UmodeOp {
+    /// Do nothing.
+    Nop = 1,
+    /// Say hello.
+    Hello = 2,
+}
+
+impl TryFrom<u64> for UmodeOp {
+    type Error = Error;
+
+    fn try_from(reg: u64) -> Result<UmodeOp, Error> {
+        match reg {
+            1 => Ok(UmodeOp::Nop),
+            2 => Ok(UmodeOp::Hello),
+            _ => Err(Error::RequestNotSupported),
+        }
+    }
+}
+
+/// An operation requested by the hypervisor and executed by umode.
+#[derive(Debug)]
+pub struct UmodeRequest {
+    op: UmodeOp,
+    in_addr: Option<u64>,
+    in_len: usize,
+    out_addr: Option<u64>,
+    out_len: usize,
+}
+
+impl UmodeRequest {
+    /// A Nop request: do nothing.
+    pub fn nop() -> UmodeRequest {
+        UmodeRequest {
+            op: UmodeOp::Nop,
+            in_addr: None,
+            in_len: 0,
+            out_addr: None,
+            out_len: 0,
+        }
+    }
+
+    /// Hello World.
+    pub fn hello() -> UmodeRequest {
+        UmodeRequest {
+            op: UmodeOp::Hello,
+            in_addr: None,
+            in_len: 0,
+            out_addr: None,
+            out_len: 0,
+        }
+    }
+
+    /// Returns the requested Operation.
+    pub fn op(&self) -> UmodeOp {
+        self.op
+    }
+}
+
+impl TryIntoRegisters for UmodeRequest {
+    fn try_from_registers(regs: &[u64]) -> Result<UmodeRequest, Error> {
+        let req = UmodeRequest {
+            op: UmodeOp::try_from(regs[0])?,
+            in_addr: if regs[1] == 0 { None } else { Some(regs[1]) },
+            in_len: regs[2] as usize,
+            out_addr: if regs[3] == 0 { None } else { Some(regs[3]) },
+            out_len: regs[4] as usize,
+        };
+        Ok(req)
+    }
+
+    fn set_registers(&self, regs: &mut [u64]) {
+        regs[0] = self.op as u64;
+        regs[1] = if let Some(val) = self.in_addr { val } else { 0 };
+        regs[2] = self.in_len as u64;
+        regs[3] = if let Some(val) = self.out_addr {
+            val
+        } else {
+            0
+        };
+        regs[4] = self.out_len as u64;
+    }
+}
+
+// HypCall: calls from umode to hypervisor.
+
+/// Calls from umode to the hypervisors.
+pub enum HypCall {
+    /// Panic and exit immediately.
+    Panic,
+    /// Print a character for debug.
+    PutChar(u8),
+    /// Return result of previous request and wait for next operation.
+    NextOp(Result<(), Error>),
```

Comment:
So all output is passed via an out pointer? Do you foresee having a scalar return value being useful for anything (e.g. to return the number of bytes written to the out pointer, or for ops that need to only return an integer)?

Reply:
Yes, I will need something like this, and that can be changed easily. I just wanted to keep this to a minimum at the moment. Getting the basic infrastructure right first.

Reply:
No changes.